samza-sql/src/main/java/org/apache/samza/sql/avro/AvroRelConverter.java

```diff
@@ -164,25 +170,69 @@ public SamzaSqlRelMessage convertToRelMessage(KV<Object, Object> samzaMessage) {
     return new SamzaSqlRelMessage(samzaMessage.getKey(), fieldNames, values);
   }
 
+  private SamzaSqlRelRecord convertToRelRecord(GenericData.Record avroRecord) {
+    List<Object> values = new ArrayList<>();
+    List<String> fieldNames = new ArrayList<>();
+    if (avroRecord != null) {
+      fieldNames.addAll(avroRecord.getSchema().getFields()
+          .stream()
+          .map(Schema.Field::name)
+          .collect(Collectors.toList()));
+      values.addAll(avroRecord.getSchema().getFields()
+          .stream()
+          .map(x -> {
+            Object obj = avroRecord.get(x.name());
+            if (obj != null && obj instanceof GenericData.Record) {
+              return convertToRelRecord((GenericData.Record) obj);
+            }
+            return obj;
+          })
+          .collect(Collectors.toList()));
+    } else {
+      String msg = "Avro Record is null";
+      LOG.error(msg);
+      throw new SamzaException(msg);
+    }
+
+    return new SamzaSqlRelRecord(fieldNames, values);
+  }
+
+  /**
+   * Convert the nested relational message to the output message.
+   */
   @Override
   public KV<Object, Object> convertToSamzaMessage(SamzaSqlRelMessage relMessage) {
     return convertToSamzaMessage(relMessage, this.avroSchema);
   }
 
   protected KV<Object, Object> convertToSamzaMessage(SamzaSqlRelMessage relMessage, Schema avroSchema) {
+    return new KV<>(relMessage.getKey(), convertToGenericRecord(relMessage.getSamzaSqlRelRecord(), avroSchema));
+  }
+
+  private GenericRecord convertToGenericRecord(SamzaSqlRelRecord relRecord, Schema avroSchema) {
     GenericRecord record = new GenericData.Record(avroSchema);
-    List<String> fieldNames = relMessage.getFieldNames();
-    List<Object> values = relMessage.getFieldValues();
+    List<String> fieldNames = relRecord.getFieldNames();
+    List<Object> values = relRecord.getFieldValues();
     for (int index = 0; index < fieldNames.size(); index++) {
       if (!fieldNames.get(index).equalsIgnoreCase(SamzaSqlRelMessage.KEY_NAME)) {
-        record.put(fieldNames.get(index), values.get(index));
+        Object obj = values.get(index);
+        String fieldName = fieldNames.get(index);
+        if (obj instanceof SamzaSqlRelRecord) {
+          record.put(fieldName,
+              convertToGenericRecord((SamzaSqlRelRecord) obj, avroSchema.getField(fieldName).schema()));
+        } else {
+          record.put(fieldName, values.get(index));
+        }
       }
     }
 
-    return new KV<>(relMessage.getKey(), record);
+    return record;
   }
 
   private Object getRelField(RelDataType relType, Object avroObj) {
+    if (avroObj instanceof GenericData.Record) {
```

Comment:
I am not entirely sure, But i don't think this will work if the schema has array of records or Map of records. I think you might need to do this conversion in AvroToRelObjConverter and for Arrays and Maps you might need to recursively call the conversion.  
Also can you add a test case for array of records and map of records if they are not already present?

Reply:
Thanks for pointing this out. I have now added support for Map, Array and Union with one nullable type.